staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -1005,7 +1005,7 @@ func (e *Store) Delete(ctx context.Context, name string, options *metav1.DeleteO
 	}
 
 	// If dry-run, then just return the object as just saved.
-	if dryrun.IsDryRun(options.DryRun) {
+	if dryrun.IsDryRun(options.DryRun) && out != nil {
```

Comment:
I didn't expect this.  Why shouldn't dry-run be honored in all cases?  It seems like even an error would be better than actually doing a delete when a dry run is requested.

Reply:
I am also confused about why we need this if statement when we also pass something about dry run into the delete statement below.

Reply:
The code *before* that condition/early-return is doing a guaranteedupdate to add the finalizer and deletion timestamp. That guaranteed update is dry-ran, which means that the object is never persisted with deletion timestamp and finalizers.
The code *after* does the delete when an immediate delete is required. Actually it does an atomic get-and-delete and returns the object as it is before being deleted (including finalizer and deletion timestamp). The dry-run delete just does a get, but in this case, it doesn't get the proper object since the guaranteed update never really happened.

The goal of this condition is just to avoid the second "get", but to return the object with the finalizer and deletion-timestamp.

Reply:
Great, can you state this in a comment in the code? It's not obvious.